Allow setting RequestContext per interaction and remove hardcoding of User Agent

arvindkrishnakumar-okta opened the pull request:
Fixes:

- #326 
- #301 

Tech Design [here](https://docs.google.com/document/d/1Tg4qjL2UcOEpQWCBS3ffPGPAEWqwUbNYyiyj26bi3ig/edit?usp=sharing).

JayNewstrom reviewed (commented):
There are lots of break API changes, do we need to bump to the next major version?

Should we store `deviceContext`, and pass it with every request? Or does sending it with the interact request hydrate the OIE pipeline for the rest of the transaction?

arvindkrishnakumar-okta commented:
> There are lots of break API changes, do we need to bump to the next major version?
> 
> Should we store `deviceContext`, and pass it with every request? Or does sending it with the interact request hydrate the OIE pipeline for the rest of the transaction?

Yes, we plan to release a next major version.

`DeviceContext` is relevant and only accepted by `interact` API endpoint in the OIE backend. We need not be sending this header unnecessarily in other APIs, I believe.

arvindkrishnakumar-okta commented:
> Should we update the sample to demonstrate this functionality as well? Seems like this is critical to make the okta platform work as expected.

From a developer's point of view, we're just exposing a way to add `DeviceContext` headers while beginning a transaction (interaction). The backend accepts these headers for `interact` endpoint and this provides a way for the developer to pass "contextual" information to Okta backend (not necessarily critical for Okta platform to work as expected, IMHO).

bdemers commented on api/src/main/java/com/okta/idx/sdk/api/client/IDXAuthenticationWrapper.java:
```diff
@@ -829,7 +823,7 @@ public AuthenticationResponse begin() {
      */
     public AuthenticationResponse beginPasswordRecovery(String token) {
```
We should probably expose the setting of the DeviceContext here too.
From the audit perspective, the user agent/IP of the client is important

bdemers replied:
Though I don't know if this method is used to start a new transaction or not. (if it isn't ignore this comment)

arvindkrishnakumar-okta replied:
Yes, it starts a new transaction with a `AuthenticationTransaction.create(..)`.

bdemers reviewed (commented):
Similar question as @JayNewstrom,  As long as the client context is only needed at the start of the flow (to figure which rules to apply).

There are probably some edge cases, where the device starts on one network and moves to another (and changing the IP), likely more common with something like a mobile device jumping between cell and wifi.

Again, I don't know if that matters at all to the underlying API. 🤷

bdemers commented:
Don't forget to bump the version ;)

bdemers reviewed (changes requested):
I was testing this out today, and ran into an issue:

```java
AuthenticationResponse authenticationResponse = idxAuthenticationWrapper.begin(
        new DeviceContext()
                .addXForwardedForHeader("1.1.1.1") // (replaced with actual IP)
                .addUserAgentHeader("actual-user-agent"));
```

Calling `begin()` calls:

```java
AuthenticationTransaction.create(client, deviceContext)
```

which calls:

```java
AuthenticationTransaction.create(client, null, null, deviceContext);
```

> **NOTE:** the `null` token (second arg)

That method has the following logic:

```java
if (token == null) {
    idxClientContext = client.interact();
} else {
    Assert.notNull(tokenType, "token type may not be null");
    idxClientContext = client.interact(token, tokenType, deviceContext);
}
```

Since `token` is null the `DeviceContext` is ignored

arvindkrishnakumar-okta commented on api/src/main/java/com/okta/idx/sdk/api/client/IDXAuthenticationWrapper.java:
```diff
@@ -844,7 +838,8 @@ public AuthenticationResponse beginPasswordRecovery(String token) {
      */
     public AuthenticationResponse beginUserActivation(String token) {
```
Yes, it is needed here. Updated PR.

arvindkrishnakumar-okta commented:
> I was testing this out today, and ran into an issue:
> 
> ```java
> AuthenticationResponse authenticationResponse = idxAuthenticationWrapper.begin(
>         new DeviceContext()
>                 .addXForwardedForHeader("1.1.1.1") // (replaced with actual IP)
>                 .addUserAgentHeader("actual-user-agent"));
> ```
> 
> Calling `begin()` calls:
> 
> ```java
> AuthenticationTransaction.create(client, deviceContext)
> ```
> 
> which calls:
> 
> ```java
> AuthenticationTransaction.create(client, null, null, deviceContext);
> ```
> 
> > **NOTE:** the `null` token (second arg)
> 
> That method has the following logic:
> 
> ```java
> if (token == null) {
>     idxClientContext = client.interact();
> } else {
>     Assert.notNull(tokenType, "token type may not be null");
>     idxClientContext = client.interact(token, tokenType, deviceContext);
> }
> ```
> 
> Since `token` is null the `DeviceContext` is ignored

Fixed and updated PR.

bdemers commented on api/src/main/java/com/okta/idx/sdk/api/model/RequestContext.java:
```diff
@@ -0,0 +1,70 @@
+     */
+    private final Map<String, String> headers = new LinkedHashMap<>();
+
+    public RequestContext addXOktaUserAgentExtendedHeader(String value) {
```
See comment in readme.

bdemers commented on README.md:
````diff
@@ -105,6 +105,17 @@ Begin Transaction with Activation token:
+
+AuthenticationResponse beginResponse = idxAuthenticationWrapper.begin(requestContext);
+```
+
````
We will need to provide guidance on the Device Token, what our recommendations are, how its used, etc.

arvindkrishnakumar-okta replied:
Sure, we can include a link to developer docs here when it is ready.

bdemers commented on samples/embedded-auth-with-sdk/src/main/java/com/okta/spring/example/helpers/Util.java:
```diff
@@ -73,4 +75,18 @@ public static void setProceedContextForPoll(final HttpSession session, final Pro
+        final RequestContext requestContext = new RequestContext();
+        // generate random str 32 char long
+        final UUID randomUUID = UUID.randomUUID();
+        final String randomString = randomUUID.toString().replaceAll("-", "");
```
Rumor has it that this field is limited to 32 characters on the server, and the need to strip the `-` are a result of that.  While it isn't a problem to strip these here.  It does add more burden on the developer, many UUID implementations add the dashes in a `toString()`.  The server should just be updated to increase this field size.  If nothing else to prevent this type of confusion to the developer.

Can you report this to the correct team?

arvindkrishnakumar-okta replied:
Initiated an internal conversation with the owning team.
